### PR TITLE
 🔧 Update Docker image build

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind other

**What this PR does / why we need it**: Attempt to improve the reliability of https://testgrid.k8s.io/sig-cluster-lifecycle-image-pushes#post-cluster-api-provider-azure-push-images. Updates image building to match the current CAPI process and leverage experimental docker / docker build kit.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Update Docker image build
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -23,20 +25,30 @@ ENV GOPROXY=$goproxy
 # Copy the Go Modules manifests
 COPY go.mod go.mod
 COPY go.sum go.sum
+
 # Cache deps before building and copying source so that we don't need to re-download as much
 # and so that source changes don't invalidate our downloaded layer
 RUN go mod download
```
Comment (cpuguy83) [inline Dockerfile:31]: Probably worth caching /go/pkg/mod as well, and mounting the go build cache along with this... not sure if `go mod download` would be building any of the mods...

File: Dockerfile [hunk 2]
```diff
@@ -23,20 +25,30 @@ ENV GOPROXY=$goproxy
...
 
 # Copy the sources
 COPY ./ ./
 
+# Cache the go build into the the Go’s compiler cache folder so we take benefits of compiler caching across docker build calls
+RUN --mount=type=cache,target=/root/.cache/go-build \
```
Comment (cpuguy83) [inline Dockerfile:37]: If you decide to cache /go/pkg/mod, will need to do that here and anywhere else you are building stuff as well.

File: Dockerfile [hunk 3]
```diff
@@ -1,3 +1,5 @@
+# syntax=docker/dockerfile:experimental
```
Comment (cpuguy83) [inline Dockerfile:1]: Might want to specify the version here so you don't wind up with some accidental breakage in the future.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 4]
```diff
@@ -1,3 +1,5 @@
+# syntax=docker/dockerfile:1.0-experimental
```
Comment (devigned) [inline Dockerfile:1]:
```suggestion
# syntax=docker/dockerfile:1.0.0-experimental
```

**Review by devigned — CHANGES_REQUESTED** [review]

**Review by cpuguy83 — APPROVED** [review]
Nice, LGTM!

**Review by vincepri — APPROVED** [review]
/approve

**Comment by vincepri** [comment]
/lgtm
per above

**Review by devigned — APPROVED** [review]
/lgtm

**Pull request merged**